tiledb/sm/group/group_details_v1.cc

```diff
@@ -75,36 +75,27 @@ Status GroupDetailsV1::apply_pending_changes() {
 
   // Remove members first
   for (const auto& member : members_to_modify_) {
-    auto& uri = member->uri();
+    auto key = member->name_or_uri();
```

Comment:
The behavior here is poorly defined for the case where a user adds an anonymous member whose URI matches the name of a named member.

Before any changes `GroupDetails` has the following:
* members_vec_: vector of shared pointers
* members_: URI or name to shared pointer
* members_by_name_: name to shared pointer (subset of members_)

I recommend we update to have the following instead:
* members_vec_: vector of shared pointers
* named_members_: mapping from name to shared pointer (renaming of `members_by_name_`)
* anon_members_: mapping from URI to shared pointer for group members that do not have a name

This way we could always check named members first, then check anonymous members if a named member was not found, and we will know in the case of a conflict the named member is always removed first.

Reply:
I didn't think of this case, thanks! I wonder if we should allow such anonymous members; for example if you call `tiledb_group_remove_member(ctx, group, "foo")`, will it remove the member with the name `foo`, or the member with no name and pointing to the URI `foo`?

I also wonder how common are groups with both named and unnamed members.

Reply:
I hope they aren't too common, but if we allow it, we should properly support it. Explicitly checking on `add` that there is no conflict between anonymous-uri and names would allow your current design to work. Just make sure to add in a test for this pathological case that either fails with a good error or has predictable behavior. Also, if you go the error route, make sure to update the PR description with another C-API breaking change.

Reply:
Calling `tiledb_group_(add|remove)_member` with the same member key many times now fails. I added some tests.

Reply:
> either fails with a good error or has predictable behavior

How about we do the latter and clearly specify that the latest addition or removal of a group member based on its key prevails? It allows doing more things from the user's side and is simpler to implement. Currently tests fail because removing and then adding group members has been blocked by my recent commit. And supporting this without supporting overwriting added group members is challenging to implement.